File: backend/src/main/java/dev/coldhands/pair/stairs/backend/infrastructure/mapper/CombinationMapper.java
package dev.coldhands.pair.stairs.backend.infrastructure.mapper;

import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.CombinationEntity;
import dev.coldhands.pair.stairs.core.domain.Combination;
import dev.coldhands.pair.stairs.core.domain.pairstream.PairStream;

import java.util.List;
import java.util.Set;

import static java.util.stream.Collectors.toSet;

public final class CombinationMapper {

    private CombinationMapper() {
    }

    public static Combination<PairStream> entityToCore(CombinationEntity entity) {
        final Set<PairStream> pairs = entity.getPairs().stream()
                .map(PairStreamMapper::entityToCore)
                .collect(toSet());

        return new Combination<>(pairs);
    }

    public static List<dev.coldhands.pair.stairs.backend.domain.PairStream> entityToDomain(CombinationEntity entity) {
        return entity.getPairs().stream()
                .map(PairStreamMapper::entityToInfo)
                .sorted()
                .toList();
    }
}

File: backend/src/main/kotlin/dev/coldhands/pair/stairs/backend/infrastructure/mapper/CombinationMapper.kt
package dev.coldhands.pair.stairs.backend.infrastructure.mapper

import dev.coldhands.pair.stairs.backend.domain.combination.PairStream
import dev.coldhands.pair.stairs.backend.domain.developer.DeveloperDao
import dev.coldhands.pair.stairs.backend.domain.stream.StreamDao
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.PairStreamInfo

class CombinationMapper(
    private val developerDao: DeveloperDao,
    private val streamDao: StreamDao
) {
    fun toInfo(combination: Collection<PairStream>): List<PairStreamInfo> {
        val developers = developerDao.findAll()
        val streams = streamDao.findAll()

        return combination.map { pairStream ->
            PairStreamInfo(
                developers = pairStream.developerIds.map { developerId ->
                    developers.firstOrNull { it.id == developerId }?.toInfo()
                        ?: error("bad times")
                }.sorted(),
                stream = streams.firstOrNull { it.id == pairStream.streamId }?.toInfo()
                    ?: error("bad times")
            )
        }
            .sorted()
    }
}